sweepbatcher: add mode with presigned transactions

starius opened the pull request:
In this mode sweepbatcher uses transactions provided by the Presigned helper.
Transactions are signed upon adding an input to a batch.

A single Batcher instance can handle both presigned and regular batches.
Currently presigned and non-presigned sweeps never appear in the same batch.



#### Pull Request Checklist
- [ ] Update `release_notes.md` if your PR contains major features, breaking changes or bugfixes

hieblmi commented on sweepbatcher/sweep_batch.go:
```diff
@@ -566,6 +580,54 @@ func (b *batch) addSweep(ctx context.Context, sweep *sweep) (bool, error) {
+				return false, nil
+			}
+		} else {
+			if err := b.ensurePresigned(ctx, sweep); err != nil {
```
if `err == nil` here, don't we then have to call `b.presign` for the sweep here?

starius replied:
If `ensurePresigned` returned no error, this means, it has been presigned already. No need to presign again. If not presigned, then crash the batcher (this is only possible because of a bug).

hieblmi commented on sweepbatcher/sweep_batch.go:
```diff
@@ -566,6 +580,54 @@ func (b *batch) addSweep(ctx context.Context, sweep *sweep) (bool, error) {
+	// use the same mode (presigned or regular).
+	if sweep.presigned {
+		// Ensure that all the sweeps in the batch use presigned mode.
+		for _, s := range b.sweeps {
```
instead of iterating, do you think it makes sense to add a `presigned` flag to the batch struct that we can check before adding a sweep to a batch?

hieblmi commented on sweepbatcher/presigned.go:
```diff
@@ -0,0 +1,441 @@
+		batchAmt += sweep.value
+	}
+
+	// Find when at least one sweep expires.
```
nit: "find the sweep with the earliest expiry"

starius replied:
Fixed

hieblmi commented on sweepbatcher/presigned.go:
```diff
@@ -0,0 +1,441 @@
+	}
+
+	// Find the feerate needed to get into next block. Use conf_target=2,
+	nextBlockFeerate, err := b.wallet.EstimateFeeRate(ctx, 2)
```
why is it 2? Maybe create a constant?

starius replied:
I updated this place:
```go
        // priorityConfTarget defines the confirmation target for quick
        // inclusion in a block. A value of 2, rather than 1, is used to prevent
        // overestimation caused by temporary anomalies, such as periods without
        // blocks.
        const priorityConfTarget = 2

        // Find the feerate needed to get into next block.
        nextBlockFeeRate, err := b.wallet.EstimateFeeRate(
                ctx, priorityConfTarget,
        )
```

This fee rate is only used to define a high fee rate (10x of this value and at least 100 sats/vbyte).

starius replied:
BTW the fee estimator fails with conf_target=1: https://github.com/lightninglabs/loop/issues/898
Maybe it worth making this a global constant in Loop and use it in all such places.

starius replied:
Updated the comment:
```go
        // priorityConfTarget defines the confirmation target for quick
        // inclusion in a block. A value of 2, rather than 1, is used to prevent
        // fee estimator from failing.
        // See https://github.com/lightninglabs/loop/issues/898
        const priorityConfTarget = 2
```

hieblmi reviewed (commented):
Amazing work @starius. I've left a couple of questions in my first pass.

starius commented on sweepbatcher/sweep_batch.go:
```diff
@@ -524,6 +560,54 @@ func (b *batch) addSweep(ctx context.Context, sweep *sweep) (bool, error) {
+		for _, s := range b.sweeps {
+			if !s.presigned {
+				b.log().Infof("failed to add sweep %x to the "+
+					"batch, because the batch has "+
```
Fixed

starius commented on sweepbatcher/sweep_batch.go:
```diff
@@ -524,6 +560,54 @@ func (b *batch) addSweep(ctx context.Context, sweep *sweep) (bool, error) {
+		// Ensure that all the sweeps in the batch don't use presigned.
+		for _, s := range b.sweeps {
+			if s.presigned {
+				b.log().Infof("failed to add sweep %x to the "+
```
Fixed

starius commented on sweepbatcher/sweep_batcher.go:
```diff
@@ -564,8 +628,40 @@ func (b *Batcher) Run(ctx context.Context) error {
 }
 
+// PresignSweep creates and stores presigned 1:1 transactions for the sweep.
+// This method must be called prior to AddSweep if presigned mode is enabled.
```
I updated `ensurePresigned` and SignTx interface (added argument `presignedOnly`), so `ensurePresigned` actually ensured that it is presigned, not tried to sign (become read-only). Also added a test making sure that the batcher crashes if AddSweep is called before calling PresignSweep (for a presigned sweep). (AddSweep can't return that error, since it only pushes a sweep to the event loop.)

lightninglabs-deploy commented:
@hieblmi: review reminder

lightninglabs-deploy commented:
@hieblmi: review reminder

lightninglabs-deploy commented:
@hieblmi: review reminder

hieblmi commented on sweepbatcher/sweep_batch.go:
```diff
@@ -1960,7 +2116,21 @@ func (b *batch) persist(ctx context.Context) error {
 
 // getBatchDestAddr returns the batch's destination address. If the batch
 // has already generated an address then the same one will be returned.
+// The method must not be used in presigned mode. Use getSweepsDestAddr instead.
```
should we call this method in the comment `getPresignedSweepsDestAddr` to clarify?

starius replied:
Renamed. Thanks!

hieblmi commented on sweepbatcher/presigned.go:
```diff
@@ -0,0 +1,457 @@
+
+// getSweepsDestAddr returns the destination address used by a group of sweeps.
+// The method must be used in presigned mode only.
+func (b *batch) getSweepsDestAddr(ctx context.Context,
```
could call it `getPresignedSweepsDestAddr` to make it clearer that this only makes sense in the context of presigned batches.

starius replied:
Renamed

hieblmi commented on sweepbatcher/presigned.go:
```diff
@@ -0,0 +1,457 @@
+		return fmt.Errorf("presignedHelper is not installed")
+	}
+	if len(b.sweeps) != 0 {
+		return fmt.Errorf("ensurePresigned is done when adding to an " +
```
nit: shoud be done...

starius replied:
Fixed

hieblmi commented on sweepbatcher/presigned.go:
```diff
@@ -0,0 +1,457 @@
+		ctx, tx, batchAmt, feeRate, feeRate, presignedOnly,
+	)
+	if err != nil {
+		return fmt.Errorf("failed to sign unsigned tx %v "+
```
nit: failed to load a presigned tx?

starius replied:
Fixed:

```go
                return fmt.Errorf("failed to find a presigned transaction "+
                        "for feeRate %v, txid of the template is %v: %w",
                        feeRate, tx.TxHash(), err)
```

hieblmi commented on sweepbatcher/presigned.go:
```diff
@@ -0,0 +1,457 @@
+	for _, sweep := range newSweeps {
+		batchAmt += sweep.value
+	}
+	const presignedOnly = true
```
maybe we could call this flag `loadOnly` or something similar?

starius replied:
Renamed to `loadOnly`.

hieblmi commented on sweepbatcher/presigned.go:
```diff
@@ -0,0 +1,457 @@
+)
+
+// ensurePresigned checks that we can sign a 1:1 transaction sweeping the input.
+func (b *batch) ensurePresigned(ctx context.Context, newSweeps []*sweep) error {
```
does `ensurePresigned` imply that there should be already be a transaction with given inputs that has a populated witness? I don't see a check anywhere in this method or sub-method that checks for a witness.

starius replied:
I added a check to `CheckSignedTx` which is called by this method:

```go
// Make sure all inputs of signedTx have a non-empty witness.
for _, txIn := range signedTx.TxIn {
	if len(txIn.Witness) == 0 {
		return fmt.Errorf("input %s of signed tx is not signed",
			txIn.PreviousOutPoint)
	}
}
```

hieblmi commented on sweepbatcher/presigned.go:
```diff
@@ -0,0 +1,457 @@
+		}
+
+		// Try to presign this transaction.
+		err = presigner.Presign(ctx, tx, batchAmt)
```
Just a side question: Currently the result is not yet used. Is it intended to then pass this pre-signed transaction into our store? I guess once the batch is confirmed we could clean up all the presigned fee versions.

starius replied:
`PresignHelper` is responsible for storing the transactions. SweepBatcher relies on `helper.SignTx` to be able to load transactions stored by previous `helper.Presign` and `helper.SignTx` calls.

> I guess once the batch is confirmed we could clean up all the presigned fee versions.

This is done. There is the method `CleanupTransactions` which is called by `batch.handleConf` method.

hieblmi commented on sweepbatcher/presigned.go:
```diff
@@ -0,0 +1,457 @@
+		ctx, tx, batchAmt, minRelayFee, feeRate, presignedOnly,
+	)
+	if err != nil {
+		return 0, fmt.Errorf("failed to sign tx: %w", err),
```
is it trying to sign when `presignedOnly` is passed in?

starius replied:
`presignedOnly = false` in this call, so it tries to sign interactively and if it fails, goes to the store of presigned transactions.

hieblmi commented on sweepbatcher/presigned.go:
```diff
@@ -0,0 +1,457 @@
+}
+
+// CheckSignedTx makes sure that signedTx matches the unsignedTx. It checks
+// according to criteria specified in the description of PresignedHelper.SignTx.
```
looks like the input sequence numbers aren't checked as described in the godoc of `SignTx`.

starius replied:
Sequence numbers are compared using `unsignedMap`. Also there is a test case to verify this.

hieblmi reviewed (commented):
It's great work @starius, I've left a couple of more comments, looks almost good to me.

starius commented on sweepbatcher/sweep_batcher.go:
```diff
@@ -156,6 +156,53 @@ type SignMuSig2 func(ctx context.Context, muSig2Version input.MuSig2Version,
+
+	// DestPkScript returns destination pkScript used in a presigned
+	// transaction sweeping the inputs. Returns an error, if such tx
+	// doesn't exist. If there are many such transactions, returns any
```
Yes, all of the transactions should have the same destination address, by construction. However this is not enforced by `PresignedHelper`. In theory someone can call `Presign` on transactions with the same list of inputs and different destination addresses and all of the addresses would be considered valid. We can demand `PresignedHelper` to check that address is always the same, but this would make queries heavier. (Checking in `Presign` that destination address is the same in all presigned transactions with such inputs.)

starius commented on sweepbatcher/sweep_batch.go:
```diff
@@ -587,6 +602,80 @@ func (b *batch) addSweeps(ctx context.Context, sweeps []*sweep) (bool, error) {
+				return false, nil
+			}
+
+		// If this is new batch being formed, make sure we already have
```
Fixed

starius commented:
Found a couple of edge cases. Converting to a draft until the fix is ready.

starius commented:
Fixed a couple of edge cases:

1. The primary sweep ID (UTXO) is now passed along with presigned transactions to help `PresignedHelper` identify the batch. The primary sweep is always spent when the batch is spent. Other sweeps may remain unspent if an intermediate version of the batch is confirmed.

2. If some sweeps in a batch remain unspent after the batch is confirmed, we need to handle them properly. I updated the code to prepare presigned transactions in advance for the sweeps that can remain unspent.

**Example for the second issue:**  
`swap1` is added to the batch. We presign transaction for [`swap1`].

Then `swap2` is added. At this point, we already have a presigned transaction for [`swap2`] created by `PresignSweepsGroup`. When adding `swap2`, we also generate a presigned transaction for [`swap1`, `swap2`].

Then `swap3` is added. Which transactions should we presign?
- [`swap1`, `swap2`, `swap3`] — the full transaction  
- [`swap2`, `swap3`] — needed in case `swap1` is confirmed on its own  
- [`swap3`] — already created by `PresignSweepsGroup`, needed if [`swap1`, `swap2`] is confirmed  

I've added unit tests for all possible cases.

---

Changes since previous review round:
- added commit "sweepbatcher: auto-increment sweep.ID in StoreMock"
- many changes in the last commit, [the diff is here](https://github.com/starius/loop/compare/v0.30.0-presigned-beta..presigned-readd-fixed2)

starius commented:
Rebased and updated the last commit. In the `handleConf` function, we now use the transaction from the confirmation notification to retrieve the list of inputs for presigned storage cleanup, instead of relying on `b.sweeps`. This is a precaution because `b.sweeps` should contain the same set of inputs, as non-confirmed sweeps are removed in `handleSpend`. However, using the confirmed transaction is more reliable and clearer.

Also added a check to ensure that the transaction from the confirmation notification matches the one from the spend notification. If they differ, a warning is logged.

starius commented:
Rebased

hieblmi commented on sweepbatcher/sweep_batch.go:
```diff
@@ -1814,41 +1952,92 @@ func (b *batch) handleSpend(ctx context.Context, spendTx *wire.MsgTx) error {
+			purgeList[L-1].SwapHash == newSweep.swapHash {
+
+			// Add the input to existing SweepRequest for this swap.
+			purgeList[L-1].Inputs = append(
```
is it correct that if we have 3 pruged sweeps in presigned mode A, B, C, then `purgeList` will have slices [A], [A,B], [A,B,C]?

starius replied:
`purgeList` won't have the same sweep in multiple slices. The code groups sweeps by swap hash. `purgeList` reproduces the arguments and the order of the original external `sweepbatcher.AddSweep` calls.

Check the example in `testPresigned_purging`. It adds multiple swaps, each of two sweeps. It adds the swaps to the batcher, then all endpoints go offline (form this point it relies only on presigned transactions). Then it confirmes sweeps of some swaps (sequentially from the first to some swap - because they are published in such order). The sweeps of remaining swaps are added one by one (swap by swap) back to the batcher. The first of the remaining swaps is added to new batch, then other swaps are appended to it. All the transactions for intermediate steps and to the final version were presigned.

I added the following comment:

```go
// In presigned mode we should form a SweepRequest per swap
// (i.e. per group) and keep them ordered. It should reproduce
// the arguments and the order of the original external AddSweep
// calls.
L := len(purgeList)
```

hieblmi replied:
Thanks for the explanation!

hieblmi commented on sweepbatcher/sweep_batch.go:
```diff
@@ -1814,41 +1952,92 @@ func (b *batch) handleSpend(ctx context.Context, spendTx *wire.MsgTx) error {
-						Value:    newSweep.value,
-					},
-				},
+				Inputs:   []Input{newInput},
```
so when we are in `presigned` mode and the purged list is empty we will land here the first time, right.

starius replied:
We can reach this line either in regular mode or in presigned mode in the beginning or when new swap starts.

I added a comment:

```go
} else {
		// Add the current sweep as a new element to purgeList.
		// This is possible either in regular mode or in
		// presigned mode in the beginning or on new swap.
		purgeList = append(purgeList, SweepRequest{
				SwapHash: newSweep.swapHash,
				Inputs:   []Input{newInput},
				Notifier: newSweep.notifier,
		})
}
```

hieblmi commented on sweepbatcher/sweep_batch.go:
```diff
@@ -1917,7 +2106,40 @@ func (b *batch) handleSpend(ctx context.Context, spendTx *wire.MsgTx) error {
+	}
+
+	if presigned {
+		b.Infof("Cleaning up presigned store")
```
maybe add this new functionality to the godoc of this method.

starius replied:
Done.

```go
// handleConf handles a confirmation notification. This is the final step of the
// batch. Here we signal to the batcher that this batch was completed. We also
// cleanup up presigned transactions whose primarySweepID is one of the sweeps
// that were spent and fully confirmed: such a transaction can't be broadcasted
// since it is either in a block or double-spends one of spent outputs.
func (b *batch) handleConf(ctx context.Context,
        conf *chainntnfs.TxConfirmation) error {
```

hieblmi commented on sweepbatcher/presigned.go:
```diff
@@ -0,0 +1,602 @@
+// ensurePresigned checks that we can sign a transaction sweeping the inputs of
+// this group only. This method is called when a group of sweeps of new swap is
+// being added to the batch. If the batch can be either empty or existing. This
+// is know to the caller and affects allowNonEmptyBatch flag.
```
nit: known

starius replied:
This part of the godoc was removed in the updated version.

hieblmi commented on sweepbatcher/presigned.go:
```diff
@@ -0,0 +1,602 @@
+
+// ensurePresigned checks that we can sign a transaction sweeping the inputs of
+// this group only. This method is called when a group of sweeps of new swap is
+// being added to the batch. If the batch can be either empty or existing. This
```
nit: comment seems off

starius replied:
Fixed.

```go
// ensurePresigned checks there there is a presigned transaction spending the
// inputs of this group only. If allowNonEmptyBatch is false, the batch must be
// empty.
func (b *batch) ensurePresigned(ctx context.Context, newSweeps []*sweep,
        allowNonEmptyBatch bool) error {
```

Also I factored out pure function `ensurePresigned` from the method and added a unit test for it.

hieblmi commented on sweepbatcher/presigned.go:
```diff
@@ -0,0 +1,602 @@
+}
+
+// getSweepsGroups returns groups in which sweeps were added to the batch.
+// All the sweeps are sorted by addition order and groupped by swap.
```
nit: groupped

starius replied:
Fixed

hieblmi commented on sweepbatcher/presigned.go:
```diff
@@ -0,0 +1,602 @@
+			groups = append(groups, []sweep{})
+		}
+
+		index := len(groups) - 1
```
move this to the start of the for loop and use it in the `if`.

starius replied:
Fixed.

```go
groups := [][]sweep{}
for _, s := range orderedSweeps {
		index := len(groups) - 1

		// Start new group if there are no groups or new swap starts.
		if len(groups) == 0 || s.swapHash != groups[index][0].swapHash {
				groups = append(groups, []sweep{})
				index++
		}

		groups[index] = append(groups[index], s)
}
```

hieblmi commented on sweepbatcher/presigned.go:
```diff
@@ -0,0 +1,602 @@
+			b.cfg.chainParams,
+		)
+		if err != nil {
+			return fmt.Errorf("failed to find destination address: %w", err)
```
nit: lll

starius replied:
Fixed

hieblmi reviewed (commented):
Left some comments and questions for the latest set of changes.

starius commented on sweepbatcher/sweep_batch.go:
```diff
@@ -1814,41 +1952,92 @@ func (b *batch) handleSpend(ctx context.Context, spendTx *wire.MsgTx) error {
 				notifyList = append(notifyList, sweep)
+				confirmedSweeps = append(
+					confirmedSweeps, sweep.outpoint,
+				)
```
Fixed

starius commented on sweepbatcher/sweep_batch.go:
```diff
@@ -1814,41 +1952,92 @@ func (b *batch) handleSpend(ctx context.Context, spendTx *wire.MsgTx) error {
+	// Sort sweeps by the addition order if this is presigned mode.
+	var allSweeps []sweep
+	if presigned {
+		allSweeps, err = b.getOrderedSweeps(ctx)
```
Fixed. I added a comment:

```go
// Sort sweeps by the addition order. This is important in presigned
// mode to pass them in correct order to purger (AddSweep) so the
// primary sweep is determined correctly and the presigned transaction
// is found. In regular mode the order doesn't matter, but we do it the
// same way for simplicity.
allSweeps, err = b.getOrderedSweeps(ctx)
```

starius commented on sweepbatcher/sweep_batcher.go:
```diff
@@ -589,8 +658,51 @@ func (b *Batcher) Run(ctx context.Context) error {
+	if err != nil {
+		return fmt.Errorf("failed to get nextBlockFeeRate: %w", err)
+	}
+	infof("PresignSweepsGroup. NextBlockFeeRate is %v", nextBlockFeeRate)
```
I updated the log message. I want to log the fact that we are in PresignSweepsGroup and also log the value of `nextBlockFeeRate`.

```go
infof("PresignSweepsGroup: nextBlockFeeRate is %v", nextBlockFeeRate)
```

starius commented on sweepbatcher/sweep_batcher.go:
```diff
@@ -589,8 +658,51 @@ func (b *Batcher) Run(ctx context.Context) error {
 }
 
+// PresignSweepsGroup creates and stores presigned transactions for the sweeps
+// group. This method must be called prior to AddSweep if presigned mode is
```
I updated `AddSweep` to do the check.

Added new commit: `sweepbatcher: load sweeps in AddSweep`.

`AddSweep` now prepares sweeps (loads from DB and fee rate source). It also needs `ctx` to work. Another small fix for ctx is in commit `sweepbatcher: use WithoutCancel for notifications`.

`AddSweep` now fails if not `PresignSweepsGroup` wasn't called before. I could just put whole presigning into `AddSweep`, but I want to put presigning is a different action so we can presign and only if presigning succeeds, switch to the action which makes `AddSweep` calls.

starius commented on sweepbatcher/sweep_batcher.go:
```diff
@@ -589,8 +658,51 @@ func (b *Batcher) Run(ctx context.Context) error {
+
 // AddSweep adds a sweep request to the batcher for handling. This will either
-// place the sweep in an existing batch or create a new one.
+// place the sweep in an existing batch or create a new one. In presigned mode
```
Fixed. See the above comment.

starius commented:
@hieblmi Thanks for comments! 🎉
I fixed them.

Also found a small edge case for function `constructUnsignedTx`. Fixed it in commit `sweepbatcher: fix rounding in constructUnsignedTx`.

starius commented:
Updated the last commit `sweepbatcher: add mode with presigned transactions`. Skipped `ensurePresigned` check in `AddSweep` if the sweep is fully confirmed (the presigned transaction is cleaned up by this time).

hieblmi reviewed (approved):
Really amazing work on this @starius. LGTM!

starius commented:
@hieblmi One more change. I replaces `helper.IsPresigned()` method with a field in `SweepInfo`. This makes things easier on naut side. I pushed it in a separate temp commit.

bhandras commented on sweepbatcher/sweep_batcher_presigned_test.go:
```diff
@@ -0,0 +1,1577 @@
+}
+
+// testPresigned_forgotten_presign checks that adding sweeps causes the batcher
+// to fail of the sweeps were not presigned with PresignSweepsGroup. In addition
```
nit: of the => if the

starius replied:
Fixed

bhandras commented on sweepbatcher/sweep_batcher_presigned_test.go:
```diff
@@ -0,0 +1,1577 @@
+}
+
+// TestPresigned tests presigned mode. Most sub-tests doesn't use loopdb.
+func TestPresigned(t *testing.T) {
```
Really nice tests and coverage!!

bhandras reviewed (commented):
Looks really good! My only complaint is the possible timeout overflow, otherwise LGTM!

starius commented on sweepbatcher/sweep_batch_test.go:
```diff
@@ -0,0 +1,318 @@
+		0x02, 0x19, 0x2d, 0x74, 0xd0, 0xcb, 0x94, 0x34, 0x4c, 0x95,
+		0x69, 0xc2, 0xe7, 0x79, 0x01, 0x57, 0x3d, 0x8d, 0x79, 0x03,
+		0xc3, 0xeb, 0xec, 0x3a, 0x95, 0x77, 0x24, 0x89, 0x5d, 0xca,
+		0x52, 0xc6, 0xb4}
```
Fixed. Thanks!

starius commented on sweepbatcher/sweep_batcher.go:
```diff
@@ -589,11 +594,32 @@ func (b *Batcher) Run(ctx context.Context) error {
+// and adds them to the batcher for handling. This will either place the sweep
+// in an existing batch or create a new one. The method can be called multiple
+// times, but the sweeps (including the order of them) must be the same. If
+// notifier is provided, the batcher sends back sweeping results though it.
```
Fixed

starius commented on sweepbatcher/presigned.go:
```diff
@@ -0,0 +1,625 @@
+	)
+
+	// Calculate the locktime value to use for high feerate transactions.
+	highFeeRateLockTime := uint32(timeout - timeoutThreshold)
```
I changed this to fix this edge case. Thanks! Added a test case.

```go
// Calculate the locktime value to use for high feerate transactions.
// If timeout <= timeoutThreshold, don't set LockTime (keep value 0).
var highFeeRateLockTime uint32
if timeout > timeoutThreshold {
	highFeeRateLockTime = uint32(timeout - timeoutThreshold)
}
```

starius commented on sweepbatcher/presigned.go:
```diff
@@ -0,0 +1,625 @@
+	numSweeps := len(tx.TxIn)
+	b.Infof("attempting to publish custom signed tx=%v, desiredFeerate=%v,"+
+		" signedFeeRate=%v, weight=%v, fee=%v, sweeps=%d, destAddr=%s",
+		txHash, feeRate, signedFeeRate, weight, fee, numSweeps, address)
```
Fixed.

starius commented on sweepbatcher/sweep_batch.go:
```diff
@@ -586,6 +601,84 @@ func (b *batch) addSweeps(ctx context.Context, sweeps []*sweep) (bool, error) {
+		}
+	} else {
+		// Ensure that all the sweeps in the batch don't use presigned.
+		for _, s := range b.sweeps {
```
Removed that `else` branch.

starius commented:
@bhandras I pushed two commits to squash them later. They have the fixes for your comments.
Please take another look!

bhandras reviewed (approved):
Really nice additions @starius 🥇 LGTM! 🎉
